sql/sql_delete.cc

```diff
@@ -584,7 +584,7 @@ bool mysql_delete(THD *thd, TABLE_LIST *table_list, COND *conds,
     {
       DBUG_PRINT("info", ("Trying direct delete"));
       if (select && select->cond &&
-          (select->cond->used_tables() == table->map))
+          (select->cond->used_tables() & table->map))
```

Comment:
Normally direct update/delete should be disabled (i.e. direct_update_rows_init/direct_delete_rows_init returns TRUE) when a condition is skipped (i.e. open_item_func returns ER_SPIDER_COND_SKIP_NUM). However, a stored function cond is never pushed to ha_spider::condition, causing the condition to be ignored. This is because Item_func_sp::fix_fields() adds RAND_TABLE_BIT to Item_func_sp::used_tables_cache, then Item_func_sp::used_tables() never EQUALS table->map.

I'm not sure if there's a good reason why here `==` is originally used instead of `&` like elsewhere. Let me know if this change may affect any behaviors.

Reply:
I think `==` is generally correct, you cannot push a condition if it uses other tables. `RAND_TABLE_BIT` is an exception, this condition could be `select->cond->used_tables() & ~RAND_TABLE_BIT == table->map`.

but generally, I'd say, then if `cond_push` is skipped or fails, there should be no attempt to use direct update at all. Like
```diff
-     if (select && select->cond &&
-           (select->cond->used_tables() == table->map))
+     bool use_direct_updates= !select || !select->cond;
+     if (!use_direct_updates &&
+          (select->cond->used_tables() & ~RAND_TABLE_BIT == table->map))
      {
        DBUG_ASSERT(!table->file->pushed_cond);
        if (!table->file->cond_push(select->cond))
+       {
+         use_direct_updates= true;
          table->file->pushed_cond= select->cond;
+       }
      }
-     if (!table->file->direct_delete_rows_init())
+     if (use_direct_updates && !table->file->direct_delete_rows_init())
```